mplhep/plot.py

```diff
@@ -20,7 +20,7 @@
 
 def histplot(h, bins, weights=None, yerr=None, variances=None,

```

Comment:
It would be ideal if you could change the names of the functions and the defaults at the same time. In other words, keep the interface of histplot as it is, but deprecate it. Add a new function `hist` without the `edges` keyword, because we always want the edges. We never want to draw without them.

Reply:
In fact, the interface of `hist` should get a few more changes. Interfaces should be minimally sufficient. Redundancies must be avoided.

*  This means `variances` should be removed, because it seems redundant with `yerr`. 
* `weights`should be removed, it makes no sense for prebinned histograms
* Perhaps: Remove `stack` and automatically draw a stacked histogram when h is a collection of arrays
* Remove `densitymode`, just use `density` like `pyplot.hist` does (if the user wants the other kind of density, which is not a real density anyway so the name is wrong, they have to calculate it themselves)
* Remove `binticks`, there are already ways to manipulate ticks in mpl

Reply:
I was hoping to work on this interface as part of Hist (probably during the Summer with a student, but maybe a bit before). I would like the new interface to support histogram objects (boost-histograms and hist histograms), probably by making edges optional, and pulling them from the first object if not given. In that case, just keeping the current interface patched up and "good enough" and the replacing it with a new one later might be the best course of action.

Are you supporting Python 2? If not, you could also make the new interface keyword only after bins, which would reduce the potential for errors when switching and would be safer to change in the future. `hist(h, bins=None, *, ...)`

Reply:
I want the new interface as soon as possible, so I can tell the people in our group that they can plot numpy histograms now. Agreed, the interface should work with numpy histograms and boost-histogram. So then the first argument should either be the numpy histogram tuple, or a boost-histogram or whatever. No bins argument.
`hist(h, *, ...)` with h either boost-histogram or `(n, edges)` from numpy.histogram.

Reply:
I can accommodate the numpy hist very quickly. In fact, in way it kind of already works `hep.histplot(*np.histogram())` I haven't played with boost::histogram yet, but addin a object check and extracting values/edges/variances should be straightforward.

Reply:
The nice thing in @HDembinski's suggestion is that histograms are already objects; you can't add a boost-histogram with `*h`, since that doesn't expand to a numpy histogram (you can with `*h.to_numpy()` of course, but that's not ideal). Then you don't need to figure out what to do with the bin argument. I think this is sort of how Aghast and Uproot handle numpy histograms; they are always passed around and recognized as a tuple. @jpivarski could correct me if I'm wrong.

Reply:
@HDembinski I think it's okay to wait a bit (if we do need to) to make sure we get the interface right. If @andrzejnovak likes your suggestion of h being a tuple for numpy, that sounds okay to me.

Reply:
I always intended that this should take objects in the end, so I am happy with that change. I do think it should be possible to plot a hist manually without constructing a tuple or an object first.
So I would prefer to solve this with `hist(h, bins=None, *, ...)` or `hist(h, *, bins=None, ...)` if need be, but imho one should be able to call `hist([1,2,1], [0,1,2,3])` rather than `hist([1,2,1], bins=[0,1,2,3])` Keeping in mind I think this should be somewhat compatible to usual matplotlib functions and `plt.hist` allows for equivalent use.

@HDembinski would you find this acceptable?

Reply:
One concern I have about taking objects is that any library this accommodates is then a dependency, which is imho kind of a weird position. I almost like the idea better of forcing any object through aghast and checking if one can extract the info by calling `h/edges/variances` on it.

Maybe there's a better solution I don't know about.

Reply:
This library should not depend on any histogram libraries. Hist may depend on this one, but not the other way around! What you can do, though, is to specify an interface. It could be something like `.axes.edges`, or something like `__edges__` (it's technically disallowed to use custom double underscore names, though). If the attribute is there, use it; if it's not, then treat it like a normal array.

Reply:
You need to depend on boost-histogram to support boost-histograms, since Python is so dynamic. For example, to support boost-histogram, you could check whether the method `.to_numpy()` exists and call it to get the usual numpy.histogram tuple.

@andrzejnovak Since you opened another PR with the monkey-patching idea (which was useful to understand better what were you want to go, thanks), I added a more detailed discussion on the two options that I see for the interface there. https://github.com/scikit-hep/mplhep/pull/100

Reply:
> @HDembinski I think it's okay to wait a bit (if we do need to) to make sure we get the interface right. If @andrzejnovak likes your suggestion of h being a tuple for numpy, that sounds okay to me.

From the Zen of Python:
"Now is better than never.
Although never is often better than *right* now."

I propose we discuss and implement the new interface in #100, which is new, so anything we do there does not have to be backward compatible with the existing `histplot` function.